File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.cpp
class Solution {
public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        vector<int> odd_positions = {};
        for (int ind = 0; ind < nums.size(); ind++){
            if (nums[ind] % 2 == 1){
                odd_positions.push_back(ind);
            }
        }

        if (odd_positions.size() < k) return 0;

        // Main subarray will have exactly k odds inside
        // Can shift start and end to include extra evens
        int total = 0;
        for (int first_odd = 0; first_odd <= odd_positions.size() - k; first_odd++){
            int prev_cutoff = (first_odd > 0) ? odd_positions[first_odd - 1] : -1;
            int next_cutoff =  nums.size();
            if (first_odd + k <= odd_positions.size() - 1){
                next_cutoff = odd_positions[first_odd + k];
            }

            int possible_starts = odd_positions[first_odd] - prev_cutoff;
            int possible_ends = next_cutoff - odd_positions[first_odd + k - 1];
            total += possible_starts * possible_ends;
        }

        return total;
    }
};

File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        inds_odd = []
        for ind in range(len(nums)):
            if nums[ind] % 2 == 1:
                inds_odd.append(ind)
        
        # [0], [1], [3], 4
        # len(inds_odd) = 4 k = 3 --> 0, 1
        num_valid = 0
        for i in range(len(inds_odd) - k + 1):
            ind_of_prev_odd = (i == 0) and -1 or inds_odd[i - 1]
            ind_of_next_odd = (i + k == len(inds_odd)) and len(nums) or inds_odd[i + k]
            
            valid_start = inds_odd[i] - ind_of_prev_odd
            valid_end = ind_of_next_odd - inds_odd[i + k - 1]
            num_valid += valid_start * valid_end

        return num_valid